### PR TITLE
ENG-8844 Added LICENSE to SystemInformation sysproc

### PR DESCRIPTION
**Pull request by HenryZhu123**

### CODE REVIEW

File: src/frontend/org/voltdb/SystemInformationAgent.java [hunk 1]
```diff
@@ -153,4 +159,22 @@ private void collectSystemInformationDeployment(PendingOpsRequest psr)
...
+            sendErrorResponse(psr.c, ClientResponse.GRACEFUL_FAILURE,
+                    "Unable to collect LICENSE information for @SystemInformation", psr.clientData);
+            return;
+        }
+        psr.aggregateTables = new VoltTable[1];
+        psr.aggregateTables[0] = result;
+        try {
+            sendClientResponse(psr);
+        } catch (Exception e) {
+            VoltDB.crashLocalVoltDB("Unable to return PARTITIONCOUNT to client", true, e);
```
Comment (JadenW) [inline src/frontend/org/voltdb/SystemInformationAgent.java]: The error seems irrelevant

File: src/frontend/org/voltdb/sysprocs/SystemInformation.java [hunk 2]
```diff
@@ -572,4 +597,62 @@ static private String addEscapes(String name)
...
+
+    static public VoltTable populateLicenseProperties()
+    {
+        VoltTable results = new VoltTable(clusterInfoSchema);
+        SimpleDateFormat sdf = new SimpleDateFormat("EEE MMM d, yyyy");
+
+        results.addRow("PERMIT_VERSION", Integer.toString(VoltDB.instance().getLicenseApi().getVersion()));
+        results.addRow("PERMIT_SCHEME", Integer.toString(VoltDB.instance().getLicenseApi().getScheme()));
+
+        if(VoltDB.instance().getLicenseApi().getLicenseType() != null)
```
Comment (JadenW) [inline src/frontend/org/voltdb/sysprocs/SystemInformation.java]: VoltDB.instance() is static function call to get the voltdb singleton which is relative expensive. You can cache it locally for access all the needed LicensApi related field.

File: src/frontend/org/voltdb/SystemInformationAgent.java [hunk 3]
```diff
@@ -65,6 +65,12 @@ protected void collectStatsImpl(Connection c, long clientHandle, OpsSelector sel
             collectSystemInformationDeployment(psr);
             return;
         }
+        if (subselector.equalsIgnoreCase("LICENSE")) {
+            collectSystemInformationLicense(psr);
```
Comment (JadenW) [inline src/frontend/org/voltdb/SystemInformationAgent.java]: Is License guaranteed to be global (i.e. same on all hosts)?  I am concerned that mix licenses with different set of permit features can be used.

Comment (manjujames) [reply src/frontend/org/voltdb/SystemInformationAgent.java]: There is no guarantee that the licenses match. The ticket says - 
"For the time being, there is no validation that the licenses within a cluster match. So until validation (and whatever technique is used to determine the authoritative cluster-wide license ) the sysproc will return data based on the license used by the node that processes the  sysproc."

File: src/frontend/org/voltdb/SystemInformationAgent.java [hunk 4]
```diff
@@ -86,7 +92,7 @@ private String parseParamsForSystemInformation(ParameterSet params, JSONObject o
                     first;
             }
             subselector = (String)first;
-            if (!(subselector.equalsIgnoreCase("OVERVIEW") || subselector.equalsIgnoreCase("DEPLOYMENT"))) {
+            if (!(subselector.equalsIgnoreCase("OVERVIEW") || subselector.equalsIgnoreCase("DEPLOYMENT") || subselector.equalsIgnoreCase("LICENSE"))) {
```
Comment (n3world) [inline src/frontend/org/voltdb/SystemInformationAgent.java]: It might be worth using a sorted set which uses a case insensitive comparison to hold the list of valid selectors.
```
private static final Set<Sting> VALID_SUBSELECTORS = ImmutableSortedSet.orderedBy(String.CASE_INSENSITIVE_ORDER).add("OVERVIEW", "DEPLOYMENT", "LICENSE").build();
...
if (!VALID_SUBSELECTORS.contains(subselector)) { ...
```

Comment (manjujames) [inline src/frontend/org/voltdb/SystemInformationAgent.java]: This might be good time to change this code to use an enum, instead of hardcoded Strings.

File: src/frontend/org/voltdb/SystemInformationAgent.java [hunk 5]
```diff
@@ -153,4 +159,22 @@ private void collectSystemInformationDeployment(PendingOpsRequest psr)
             VoltDB.crashLocalVoltDB("Unable to return PARTITIONCOUNT to client", true, e);
         }
     }
+
+    private void collectSystemInformationLicense(PendingOpsRequest psr)
```
Comment (manjujames) [inline src/frontend/org/voltdb/SystemInformationAgent.java]: This and the existing collectSystemInformationDeployment methods are same except for the call to SystemInformation. We should probably make a common method that takes in a parameter specifying which method in SystemInformation to call.

File: src/frontend/org/voltdb/SystemInformationAgent.java (hunk 3 above)
Comment (n3world) [reply src/frontend/org/voltdb/SystemInformationAgent.java]: However, this is not what this code is doing. What it is doing is sending the request for the license to all partitions and then comparing the resulting tables and if they are not the same returning an error message

File: src/frontend/org/voltdb/licensetool/LicenseApi.java [hunk 6]
```diff
@@ -46,4 +46,11 @@
     public String getSignature();
     public String getLicenseType();
     public boolean isUnrestricted();
+    public String getIssuerCompany();
+    public String getIssuerUrl();
```
Comment (zeemanhe) [inline src/frontend/org/voltdb/licensetool/LicenseApi.java]: Instead of one method for getting each license member, you could get license object.
